utils/smart_resource.py

```diff
@@ -0,0 +1,301 @@
+# -*- coding: utf-8 -*-
+import re
+
+from cfme.fixtures import pytest_selenium as sel
+from cfme.web_ui import form_buttons, flash
+
+
+class Unknown(object):
+    pass
+
+
+class Impossible(Unknown):
+    def __repr__(self):
+        return "Impossible"
+
+
+def _create_getter_for(key):
+    def getter(self):
+        if key in self.__updates__:
+            return self.__updates__.get(key, None)
+        else:
+            if self.__data__[key] is Unknown:
+                self._pull()
+            return self.__data__[key]
+    return getter
+
+
+def _create_setter_for(key):
+    def setter(self, value):
+        self.__updates__[key] = value
+    return setter
+
+
+class _smart_resource_meta(type):
+    def __new__(cls, name, bases, attrs):
+        super_new = super(_smart_resource_meta, cls).__new__
+        # Also ensure initialization is only performed for subclasses of Model
+        # (excluding Model class itself).
+        parents = [b for b in bases if isinstance(b, _smart_resource_meta)]
+        if not parents:
+            return super_new(cls, name, bases, attrs)
+
+        new_attrs = {"__fields__": {}}
+        # Retrieve all of the fields inherited (properties are generated but dict not)
+        for base in bases:
+            if issubclass(base, SmartResource) and base is not SmartResource:
+                for field_name, field in base.__fields__.iteritems():
+                    new_attrs["__fields__"][field_name] = field
+        # And generate new fields
+        for key, value in attrs.iteritems():
+            if not isinstance(value, Field):
+                new_attrs[key] = value
+                continue
+            # Store fields in this hidden key
+            new_attrs["__fields__"][key] = value
+            new_attrs[key] = property(
+                fget=_create_getter_for(key),
+                fset=_create_setter_for(key) if value.can_be_set else None)
```

Comment:
You could probably make a [descriptor class](https://docs.python.org/2/reference/datamodel.html#implementing-descriptors) and tighten this up a little: `new_attrs[key] = _smart_resource_property(self, key)`